server/storage/backend/backend.go

```diff
@@ -300,11 +303,33 @@ func (b *backend) ConcurrentReadTx() ReadTx {
 	// concurrentReadTx is not supposed to write to its txReadBuffer
 	return &concurrentReadTx{
 		baseReadTx: baseReadTx{
-			buf:     *buf,
-			txMu:    b.readTx.txMu,
-			tx:      b.readTx.tx,
-			buckets: b.readTx.buckets,
-			txWg:    b.readTx.txWg,
+			buf:          buf,
+			bufferCopied: true,
+			mu:           new(sync.RWMutex),
+			txMu:         b.readTx.txMu,
+			tx:           b.readTx.tx,
+			buckets:      b.readTx.buckets,
+			txWg:         b.readTx.txWg,
+		},
+	}
+}
+
+func (b *backend) ConcurrentReadTxNoCopy() ReadTx {
+	b.readTx.RLock()
+	defer b.readTx.RUnlock()
+
+	b.readTx.txWg.Add(1)
+
+	// no copy b.readTx.buf
+	return &concurrentReadTx{
+		baseReadTx: baseReadTx{
+			buf:          b.readTx.buf,
```

Comment:
if you use b.readTx.buf here, it might be the case that read operations can get the new data from later write operations. Can you explain a little bit here how this will not happen?

Reply:
```
func (tr *storeTxnRead) rangeKeys(ctx context.Context, key, end []byte, curRev int64, ro RangeOptions) (*RangeResult, error) {
    ...
    revpairs, total := tr.s.kvindex.Revisions(key, end, rev, int(ro.Limit))
    ...
    _, vs = tr.tx.UnsafeRangeWithLock(schema.Key, revBytes, nil, 0, tr.s.b)
    ...
}

```
If we can get "revBytes" from kvindex in Revisions function, we can reasonably conclude that "revBytes" must be put before the read txn.
In key bucket for bolt db, "revBytes" will not be updated, just be put or deleted.
So we will not get the new data from read buffer because requested key "revBytes" must be put before the read txn.